### PR TITLE
[FLOC-1576] Repair missing filesystems

### PR DESCRIPTION
**Pull request by itamarst**
As preliminary to breaking up creation into atomic steps, and also to recover from failures mid-way through creation, create a filesystem on an attached block device if it is missing one. Previously we would try to mount it, fail, repeat forever.

http://build.clusterhq.com/boxes-flocker?branch=repair-missing-filesystem-FLOC-1576

[Review on Reviewable](https://reviewable.io/reviews/clusterhq/flocker/2157)

### CODE REVIEW

File: flocker/node/agents/blockdevice.py [hunk 1]
```diff
@@ -61,10 +61,17 @@
 class DatasetStates(Names):
     """
     States that a ``Dataset`` can be in.
+
     """
+    # Exists, but not attached:
```
Comment (sarum90) [inline flocker/node/agents/blockdevice.py]: :grey_question: nit, varied use of colons.

File: flocker/node/agents/blockdevice.py [hunk 2]
```diff
@@ -61,10 +61,17 @@
...
     NON_MANIFEST = NamedConstant()
+    # Attached to this node but no filesystem
+    ATTACHED_NO_FILESYSTEM = NamedConstant()
+    # Attached to this node, has filesystem
     ATTACHED = NamedConstant()
+    # Mounted on this node:
     MOUNTED = NamedConstant()
+    # XXX seems pointless might delete at some point
```
Comment (sarum90) [inline flocker/node/agents/blockdevice.py]: Given the docstring of this class it seems that this should exist. Perhaps that docstring should be updated when this is removed. (States that a Dataset with an existing volume can be in or somesuch).

File: flocker/node/agents/blockdevice.py [hunk 3]
```diff
@@ -220,13 +228,6 @@ def __init__(self, blockdevice):
     u"mounted.",
 )
 
-DEVICE_PATH = Field(
```
Comment (sarum90) [inline flocker/node/agents/blockdevice.py]: :+1:  Nice catch!

File: flocker/node/agents/test/test_blockdevice.py [hunk 4]
```diff
@@ -1088,9 +1152,14 @@ class DiscoverVolumesMethod(Names):
...
 @implementer(ILocalState)
 class FakeBlockDeviceDeployerLocalState(PClass):
+    """
+    This class is a terrible idea and should be thrown out ASAP to be
+    replaced by the real thing.
+    """
     node_state = field(type=NodeState, mandatory=True)
     nonmanifest_datasets = field(type=NonManifestDatasets, mandatory=True)
     volumes = pvector_field(BlockDeviceVolume)
+    discovered_datasets = pmap_field(UUID, DiscoveredDataset)
```
Comment (sarum90) [inline flocker/node/agents/test/test_blockdevice.py]: Unfortunately the real implementation now doesn't have this member. I think it only has "datasets".

File: flocker/node/agents/test/test_blockdevice.py [hunk 5]
```diff
@@ -1088,9 +1152,14 @@ class DiscoverVolumesMethod(Names):
 
 @implementer(ILocalState)
 class FakeBlockDeviceDeployerLocalState(PClass):
+    """
+    This class is a terrible idea and should be thrown out ASAP to be
+    replaced by the real thing.
```
Comment (sarum90) [inline flocker/node/agents/test/test_blockdevice.py]: :grey_question: No action required. My bad for letting this come in during review. I didn't think it fully through, looked ready to be killed in the next step.

Should have required this to be a more validated fake.

Comment (itamarst) [inline flocker/node/agents/test/test_blockdevice.py]: I believe Tom indicated it'll be gone next issue in his stack, so it's OK.

File: flocker/node/agents/blockdevice.py [hunk 6]
```diff
@@ -1503,12 +1518,17 @@ def calculate_changes(self, configuration, cluster_state, local_state):
         ))
         mounts = list(self._calculate_mounts(
             local_node_state.devices, local_node_state.paths,
-            configured_manifestations, local_state.volumes
+            configured_manifestations, local_state.volumes,
+            local_state.discovered_datasets,
```
Comment (sarum90) [inline flocker/node/agents/blockdevice.py]: Does "discovered_datasets" exist as a parameter? Or should this just be "datasets"?

In summary, personal lesson learned about letting non-verified fakes into the codebase.

File: flocker/node/agents/blockdevice.py (hunk 2 above)
Comment (tomprince) [inline flocker/node/agents/blockdevice.py]: This will be used in https://clusterhq.atlassian.net/browse/FLOC-3386

**Comment by sarum90** [comment]
Definitely my bad on letting that Fake come in, and I think the failing test relates to that. Specifically the real implementation does not have the discovered_datasets field that the fake has.

New work looks great though! And this new break down of responsibilities I think makes each of the parts easier to reason about and test.

Resubmit once you've addressed the failing tests, I'll try to be quick to review the next iteration.

**Comment by itamarst** [comment]
As mentioned elsewhere, end-to-end testing plan is to do FLOC-1771 (get rid of compound create) at which point this is automatically tested by the regular acceptance tests. So possibly this should not be merged until 1771 stacked on this passes acceptance tests?

**Comment by sarum90** [comment]
Looks good!

Verify that the one aws failure is just a flake, I couldn't be sure by glancing at it personally.

Otherwise this looks ready to merge.

**Comment by tomprince** [comment]
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/clusterhq/flocker/2157)

**Comment by itamarst** [comment]
Just going to merge this, followup 1771 is mostly working, Swisscom has had decent success with this repairing issues.

**Pull request merged**